cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,688 @@
+
+        // Create a bloom filter view.
+        cuco::bloom_filter_ref<key_type,
+                               cuco::extent<std::size_t>,
+                               cuco::thread_scope_thread,
+                               policy_type>
+          filter{reinterpret_cast<word_type*>(filter_span[filter_idx].data()),
+                 num_filter_blocks,
+                 {},  // Thread scope as the same literal is being searched across different bitsets
+                      // per thread
+                 {}};  // Arrow policy with cudf::hashing::detail::XXHash_64 seeded with 0 for Arrow
+                       // compatibility
+
+        // If int96_timestamp type, convert literal to string_view and query bloom
+        // filter
+        if constexpr (cuda::std::is_same_v<T, cudf::string_view> and
+                      IS_INT96_TIMESTAMP == is_int96_timestamp::YES) {
+          auto const int128_key = static_cast<__int128_t>(d_scalar.value<int64_t>());
+          cudf::string_view probe_key{reinterpret_cast<char const*>(&int128_key), 12};
+          return filter.contains(probe_key);
+        } else {
+          // Query the bloom filter and store results
+          return filter.contains(d_scalar.value<T>());
+        }
+      });
+
+    return std::make_unique<cudf::column>(cudf::data_type{cudf::type_id::BOOL8},
+                                          static_cast<cudf::size_type>(total_row_groups),
+                                          std::move(results),
+                                          rmm::device_buffer{},
+                                          0);
+  }
+
+  // Creates device columns from bloom filter membership
+  template <typename T>
+  std::unique_ptr<cudf::column> operator()(cudf::size_type equality_col_idx,
+                                           cudf::data_type dtype,
+                                           ast::literal* const literal,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::device_async_resource_ref mr) const
+  {
+    // For INT96 timestamps, use cudf::string_view type and set is_int96_timestamp to YES
+    if constexpr (cudf::is_timestamp<T>()) {
+      if (parquet_types[equality_col_idx] == Type::INT96) {
+        // For INT96 timestamps, use cudf::string_view type and set is_int96_timestamp to YES
+        return query_bloom_filter<cudf::string_view, bloom_filter_caster::is_int96_timestamp::YES>(
+          equality_col_idx, dtype, literal, stream, mr);
+      }
+    }
+
+    // For all other cases
+    return query_bloom_filter<T>(equality_col_idx, dtype, literal, stream, mr);
+  }
+};
+
+/**
+ * @brief Collects lists of equality predicate literals in the AST expression, one list per input
+ * table column. This is used in row group filtering based on bloom filters.
+ */
+class equality_literals_collector : public ast::detail::expression_transformer {
```

Comment:
No need for an ast::tree in this expression converter as we only visit and collect literals for `col == lit` expressions.